### PR TITLE
get offset of symbol+subsymbol

### PR DESCRIPTION
**Pull request by scorpionzezz**
This patch is to get the symbol + sub-symbol offset from the rekall profile.

### CODE REVIEW

File: libvmi/os/windows/core.c [hunk 1]
```diff
@@ -762,6 +762,7 @@ windows_init(
     /* Need to provide this functions so that find_page_mode will work */
     os_interface = safe_malloc(sizeof(struct os_interface));
     bzero(os_interface, sizeof(struct os_interface));
+    os_interface->os_get_symbol_offset = NULL;
```
Comment (tklengyel) [inline libvmi/os/windows/core.c]: Any particular reason why this is skipped for Windows?

File: libvmi/accessors.c [hunk 2]
```diff
@@ -141,6 +141,16 @@ vmi_get_offset(
     return vmi->os_interface->os_get_offset(vmi, offset_name);
 }
 
+status_t
+vmi_get_kernel_offset(
+    vmi_instance_t vmi, 
+    const char* symbol, 
+    const char* subsymbol,
```
Comment (tklengyel) [inline libvmi/accessors.c]: I'm not particularly happy with this API. For internal use in LibVMI it's fine but it is really not intuitive to be publicly supported. Ie., when you don't specify a subsymbol it is looking for a kernel function / static variable, whereas with a subsymbol it looks for an offset in struct. These should really be two separate functions on the public side, as "vmi_get_kernel_offset" doesn't describe it accurately.

**Comment by scorpionzezz** [comment]
Hey there, thanks for your feedback :)

I did fix it, and gave the appropriate name and also implemented it for windows. May you take a look it again please?

**Comment by tklengyel** [comment]
This is ok to test.

**Comment by tklengyel** [comment]
Retest this please.

**Comment by tklengyel** [comment]
@smaresca @bdpayne @awsaba looks fine to me but needs another +1 because of the public API addition

File: libvmi/libvmi.h [hunk 3]
```diff
@@ -1731,6 +1731,34 @@ uint64_t vmi_get_offset(
...
+ * @param[in] subsymbol String name for desired subsymbol
+ * @param[out] offset of subsymbol in struct (symbol)
+ *
+ * @return VMI_SUCCESS or VMI_FAILURE
+ */
+status_t vmi_get_kernel_struct_offset(
+    vmi_instance_t vmi,
+    const char* struct_name,
+    const char* subsymbol,
+    addr_t *rva);
```
Comment (smaresca) [inline libvmi/libvmi.h]: Parameters named subsymbol and rva seem strange in the context of this function. I understand that they originate at a Rekall layer, but for public exposure in the LibVMI API, please keep them as generic as possible.

File: libvmi/libvmi.h [hunk 4]
```diff
@@ -1731,6 +1731,34 @@ uint64_t vmi_get_offset(
...
+
+/**
+ * Get the memory offset of a kernel function or a symbol
+ * @param[in] vmi LibVMI instance
+ * @param[in] symbol name for the desired symbol
+ * @param[out] offset of the kernel function or the symbol
+ *
+ * @return VMI_SUCCESS or VMI_FAILURE
+ */
+status_t vmi_get_constant_offset(
```
Comment (smaresca) [inline libvmi/libvmi.h]: Would this be more accurate to describe as returning a va of a kernel symbol? If so, could this instead be worked into the vmi_translate_ksym2v (e.g., if symbol not found elsewhere and rekall profile exists, call vmi->os_interface->os_get_constant_offset(..) ?)

File: libvmi/libvmi.h (hunk 3 above)
Comment (tklengyel) [inline libvmi/libvmi.h]: Any suggestions what to name them?

File: libvmi/libvmi.h (hunk 4 above)
Comment (tklengyel) [inline libvmi/libvmi.h]: This would make sense actually, +1

File: libvmi/libvmi.h (hunk 3 above)
Comment (scorpionzezz) [inline libvmi/libvmi.h]: How about I change rva to addr, to make it more generic?

File: libvmi/libvmi.h (hunk 4 above)
Comment (scorpionzezz) [inline libvmi/libvmi.h]: Ah yes, this make sense, so I remove the vmi_get_constant_offset in my latest commit.

File: libvmi/os/linux/core.c [hunk 5]
```diff
@@ -433,6 +434,16 @@ void linux_read_config_ghashtable_entries(char* key, gpointer value,
     _done: return;
 }
 
+status_t linux_get_kernel_struct_offset(vmi_instance_t vmi, const char* symbol, const char* subsymbol, addr_t *addr){
+    status_t ret = VMI_SUCCESS;
+    linux_instance_t linux_instance = vmi->os_data;
+    if (VMI_FAILURE == rekall_profile_symbol_to_rva(linux_instance->rekall_profile,symbol,subsymbol,addr)){
```
Comment (tklengyel) [inline libvmi/os/linux/core.c]: You can just return the call to rekall_profile_symbol_to_rva directly, no need, for the local variable ret.

File: libvmi/os/windows/core.c [hunk 6]
```diff
@@ -414,6 +414,16 @@ get_kpgd_method0(
     return VMI_FAILURE;
 }
 
+status_t windows_get_kernel_struct_offset(vmi_instance_t vmi, const char* symbol, const char* subsymbol, addr_t *addr){
+    status_t ret = VMI_SUCCESS;
+    windows_instance_t windows = vmi->os_data;
+    if (VMI_FAILURE == rekall_profile_symbol_to_rva(windows->rekall_profile,symbol,subsymbol,addr)){
```
Comment (tklengyel) [inline libvmi/os/windows/core.c]: You can just return the call to rekall_profile_symbol_to_rva directly, no need, for the local variable ret.

**Comment by tklengyel** [comment]
Retest this please.

**Comment by tklengyel** [comment]
Retest this please.

File: libvmi/os/linux/core.c (hunk 5 above)
Comment (scorpionzezz) [inline libvmi/os/linux/core.c]: okay, I changed it in my latest commit

File: libvmi/os/windows/core.c (hunk 6 above)
Comment (scorpionzezz) [inline libvmi/os/windows/core.c]: okay, I changed it in my latest commit

File: libvmi/libvmi.h (hunk 3 above)
Comment (smaresca) [inline libvmi/libvmi.h]: rva to addr makes sense sure. Please use 'member' insted of subsymbol as well

**Comment by scorpionzezz** [comment]
done, changed subsymbol -> member

**Comment by smaresca** [comment]
Thanks @scorpionzezz  !

**Comment by scorpionzezz** [comment]
no problem, thanks @smaresca  and @tklengyel  for developing this amazing tool :)

**Pull request merged**